### PR TITLE
CD-85212 Modify PullRequest Plugin to cycle through multiple credentials

### PR DESCRIPTION
**Pull request by revathi-murali**
## JIRA

* [Issue Link](https://coupadev.atlassian.net/browse/CD-85212)

## Code Reviewers
- [x] @tjackiw 
- [x] @edk

## Summary of change
-  The changes made here are used by the [ghprb plugin](https://github.com/coupa/ghprb-plugin/pull/1)  
- Added a new matcher `IdPrefixMatcher.java` which takes in a input string(Regex) and returns true if the id of the credentials matches the regex.
- Added a new method `cyclic` in `CredentialsMatchers.java` which takes the credentials with the specified id prefix (here `cycle-ghprb-auth*`) and returns the token in a cyclic manner for making the api calls

### CODE REVIEW

File: src/main/java/com/cloudbees/plugins/credentials/matchers/IdPrefixMatcher.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+ *
+ * @since 1.5
+ */
+public class IdPrefixMatcher implements CredentialsMatcher, CredentialsMatcher.CQL {
+    /**
+     * Standardize serialization.
+     *
+     * @since 2.1.0
+     */
+    private static final long serialVersionUID = -2400504567993839126L;
```
Comment (tjackiw) [inline src/main/java/com/cloudbees/plugins/credentials/matchers/IdPrefixMatcher.java]: @revathi-murali Why is this required since it's not used anywhere?

File: src/main/java/com/cloudbees/plugins/credentials/matchers/IdPrefixMatcher.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    @Override
+    public String describe() {
+        return regex_pattern.toString();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int hashCode() {
```
Comment (tjackiw) [inline src/main/java/com/cloudbees/plugins/credentials/matchers/IdPrefixMatcher.java]: @revathi-murali This doesn't seem to be used anywhere either. Is this required for the implementation of `CredentialsMatcher`?

File: src/main/java/com/cloudbees/plugins/credentials/CredentialsMatchers.java [hunk 3]
```diff
@@ -394,6 +414,40 @@ public static CredentialsMatcher noneOf(@NonNull CredentialsMatcher... matchers)
...
+                result.add(c);
+            }
+        }
+        // If there is no matching id with the prefix, then use the credential that is selected in the UI 
+        if (result.isEmpty()){
+            return firstOrNull(credentials, idMatcher);
+        }
+        else {
+            C value = result.get(rr_value.getAndIncrement() % result.size());
+            // Loggers added for debugging and testing. Will be removed once code review is done
```
Comment (PrabakaranSubramanian) [inline src/main/java/com/cloudbees/plugins/credentials/CredentialsMatchers.java]: @revathi-murali `rr_value.getAndIncrement()` - What happens if it reaches the max integer value (Integer.MAX_VALUE), we need to reset back to 0, right? If so, can we have a check to see if it reached the maximum and then set it back to 0 or 1?

File: src/main/java/com/cloudbees/plugins/credentials/CredentialsMatchers.java [hunk 4]
```diff
@@ -394,6 +414,40 @@ public static CredentialsMatcher noneOf(@NonNull CredentialsMatcher... matchers)
...
+     * first credential that matches the supplied id matcher 
+     *
+     * @param credentials    the credentials to select from.
+     * @param prefixMatcher  the prefix matcher.
+     * @param idMatcher      the id matcher
+     * @param <C>            the type of credential.
+     * @return a matching credential or the supplied default.
+     */
+    @CheckForNull
+    public static <C extends Credentials> C cyclic(@NonNull Iterable<C> credentials,
```
Comment (tjackiw) [inline src/main/java/com/cloudbees/plugins/credentials/CredentialsMatchers.java]: @revathi-murali Let's add unit tests for this method as well to make sure that it properly cycles through different credentials.

File: src/main/java/com/cloudbees/plugins/credentials/matchers/IdPrefixMatcher.java (hunk 2 above)
Comment (revathi-murali) [inline src/main/java/com/cloudbees/plugins/credentials/matchers/IdPrefixMatcher.java]: @tjackiw Yeah it is required for the implementation of the CredentialsMatcher.

File: src/main/java/com/cloudbees/plugins/credentials/CredentialsMatchers.java (hunk 3 above)
Comment (revathi-murali) [inline src/main/java/com/cloudbees/plugins/credentials/CredentialsMatchers.java]: @PrabakaranSubramanian  Yeah even I thought the same. Will push the changes related to that.

File: src/main/java/com/cloudbees/plugins/credentials/matchers/IdPrefixMatcher.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    @NonNull
+    private final Pattern regex_pattern;
+
+    /**
+     * Constructs a new instance.
+     *
+     * @param regex_pattern the pattern to match 
+     */
+    public IdPrefixMatcher(@NonNull String regex_pattern) {
+        regex_pattern.getClass();
```
Comment (PrabakaranSubramanian) [inline src/main/java/com/cloudbees/plugins/credentials/matchers/IdPrefixMatcher.java]: @revathi-murali `regex_pattern.getClass();` Why is this needed? If it is a generic class, then may be  it helps in identifying if it is Null (as it will throw NPE), but when there is already `@NonNull` annotation and you know it is `String` class, is it still needed?

**Comment by revathi-murali** [comment]
@tjackiw @PrabakaranSubramanian Addressed the review comments. Kindly check once and update me

**Comment by edk** [comment]
lgtm.  👍   
(I can't update the checkbox since i don't have write perms to this repo).  It may not be possible from the credentials context, but it might be nice to check the `/rate_limit` endpoint to see if the limit has been hit for a particular key.  Github doesn't count the rate limit check against the rate limit.

**Comment by revathi-murali** [comment]
[lockit v4.0.0] Merged via Lockit CLI on 2017-07-11 14:25:34.290162808 +0530 IST

**Comment by revathi-murali** [comment]
@edk Yeah currently we are **not rotating** the keys **based on the hit count**. We are **just cycling the tokens which match a prefix** . But we can enhance by taking the hit count for a key in the next release or version.
